### PR TITLE
torchopt post

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+
+[^2]: For why it shouldn't, and a overview of [some of] the things `torch` users can do, see also our recent post on [extending `torch`](https://blogs.rstudio.com/ai/posts/2022-04-27-torch-outside-the-box/).
+
+[^3]: See the [Readme](https://github.com/e-sensing/torchopt) for an up-to-date listing.
+
+In this text, instead of a dry summary of optimizers included, I'm going to introduce the package by highlighting something that technically, is "merely" a utility function, but to the user, can be extremely helpful: namely, the ability to, for an arbitrary optimizer and an arbitrary test function, plot the steps taken in optimization.
+
+While it's true that I have no intent of comparing (let alone analyzing) different strategies, there is one that, to me, stands out in the list: ADAHESSIAN [@abs-2006-00719], a second-order algorithm designed to scale to large neural networks. I'm especially curious to see how it behaves as compared to L-BFGS, the second-order "classic" available from base `torch` we've had a [dedicated blog post](https://blogs.rstudio.com/ai/posts/2021-04-22-torch-for-optimization/) about last year.
```
Comment (batpigandme) [inline _posts/2022-05-18-torchopt/torchopt.Rmd:45]: If relevant, say what these stand for. (I, a naïve reader, have no clue).

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+```{r}
+flower <- function(x, y) {
+  a <- 1
+  b <- 1
+  c <- 4
+  a * torch_sqrt(torch_square(x) + torch_square(y)) + b * torch_sin(c * torch_atan2(y, x))
+}
+```
+
+Let's begin with our explorations. There'll be three short sections.
```
Comment (batpigandme) [inline _posts/2022-05-18-torchopt/torchopt.Rmd:68]:
```suggestion
Let's begin with our explorations.
```

Comment (batpigandme) [reply _posts/2022-05-18-torchopt/torchopt.Rmd:68]: Unless you're going to tell me what the sections are (which I don't think you need to), then it seems unnecessary to announce that.

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+## Why do they always say learning rate matters?
+
+True, that's a rhetorical question. But still, sometimes visualizations make for the most memorable evidence.
```
Comment (batpigandme) [inline _posts/2022-05-18-torchopt/torchopt.Rmd:72]:
```suggestion
True, it's a rhetorical question. But still, sometimes visualizations make for the most memorable evidence.
```

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+
+![Minimizing the flower function with AdamW, lr = 0.1: Successive "exploration" of petals. Steps (clockwise): 300, 700, 900, 1300.](images/petals.png)
+
+Who says you need chaos[^4] to produce a beautiful plot?
+
+[^4]: Alluding to strange attractors here. If you're interested, you might want to check out the mini-series on chaos and deep learning: [Deep attractors: Where deep learning meets chaos](https://blogs.rstudio.com/ai/posts/2020-06-24-deep-attractors/), [Time series prediction with FNN-LSTM](https://blogs.rstudio.com/ai/posts/2020-07-20-fnn-lstm/), and [FNN-VAE for noisy time series forecasting](https://blogs.rstudio.com/ai/posts/2020-07-31-fnn-vae-for-noisy-timeseries/).
+
+## A second-order optimizer for neural networks: ADAHESSIAN
+
+On to the one algorithm I'd like to check out specifically. Subsequent to just a little bit of learning-rate experimentation, I was able to arrive at an excellent result after just thirty-five steps.
```
Comment (batpigandme) [inline _posts/2022-05-18-torchopt/torchopt.Rmd:155]:
```suggestion
On to the one algorithm I'd like to check out specifically. Subsequent to a little bit of learning-rate experimentation, I was able to arrive at an excellent result after just thirty-five steps.
```

Comment (batpigandme) [reply _posts/2022-05-18-torchopt/torchopt.Rmd:155]: Since you have "just" in the following sentence, and already have a couple of qualifiers in the first, I cut it.

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+)
+```
+
+![Minimizing the flower function with ADAHESSIAN. Setup no. 2: lr = 0.3, 200 steps.](images/adahessian-steps-200.png)
+
+Like AdamW, ADAHESSIAN goes on to "explore" the petals, but it does not stray as far away from the minimum.
+
+Is this surprising? I wouldn't say it is. The argument is the same as with AdamW, above: Its algorithm has been tuned to perform well on large neural networks, not to solve a classic, hand-crafted minimization task.
+
+Now we've heard that argument twice already, it's time to verify the explicit assumption: that a classic second-order algorithm handles this better. In other words, it's time to revisit L-BFGS.
```
Comment (batpigandme) [inline _posts/2022-05-18-torchopt/torchopt.Rmd:185]: Very nice transition!

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+
+## Best of the classics: Revisiting L-BFGS
+
+To use `test_optim()` with L-BFGS, we need to take a little detour. If you've read the post on L-BFGS, you may remember that with this optimizer, it is necessary to wrap both the call to the test function and the evaluation of the gradient in a closure. (The reason being that both have to be callable several times per iteration.)
```
Comment (batpigandme) [inline _posts/2022-05-18-torchopt/torchopt.Rmd:189]: Link to the post on L-BFGS again here (I know you have it earlier, but it's convenient for your reader)

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+
+Now, seeing how L-BFGS is a very special case, and few people are likely to use `test_optim()` with it in the future, it wouldn't seem worthwhile to make that function handle different cases. For this on-off test, I simply copied and modified the code as required. The result, `test_optim_lbfgs()`, is found in the appendix.
```
Comment (batpigandme) [inline _posts/2022-05-18-torchopt/torchopt.Rmd:191]: I believe you can link to the appendix (it's an anchored header). So, I'd  add that where you mention the appendix.

**Review by batpigandme — APPROVED** [review]
Nicely done! I've added some suggestions as to where you might tighten things up a bit, but overall it reads well.

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+    self_contained: false
+    toc: true
+preview: images/petals.png
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE, fig.width = 6, fig.height = 6)
+```
+
+Pretty much from the beginnings, it has been exciting to watch the growing number of packages developing in the `torch` ecosystem. Well, not just the number. What's amazing is the variety of things people do with `torch`: extend its functionality; integrate and put to domain-specific use its low-level automatic differentiation infrastructure; port neural network architectures ... and last but not least, answer scientific questions.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2022-05-18-torchopt/torchopt.Rmd:27]: hm hm ... I'm of course tempted to put the other thing back in ;-))
... but yeah, I'm probably being pretty verbose already, asides & all ;-)

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+
+This blog post will introduce, in short and rather subjective form, one of these packages: [`torchopt`](https://github.com/e-sensing/torchopt). Before we start, one thing we should probably say a lot more often: If you'd like to publish a post on this blog, on the package you're developing or the way you employ R-language deep learning frameworks, let us know -- you're more than welcome![^1]
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2022-05-18-torchopt/torchopt.Rmd:29]: Hm yeah, I see ... I'll probably leave it there  anyway - instead of moving to the end - just to have it in focus _once_ ... but will move the link then , as you say

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+
+[^1]: For example, by creating an issue on GitHub, in either the [blog's repo](https://github.com/rstudio/ai-blog) or that of the respective software (e.g., [keras](https://github.com/rstudio/keras), [torch](https://github.com/mlverse/torch), ...)
+
+## `torchopt`
+
+`torchopt` is a package developed by [Gilberto Camara](https://gilbertocamara.org/) and colleagues at [National Institute for Space Research, Brazil](https://www.gov.br/inpe/pt-br).
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2022-05-18-torchopt/torchopt.Rmd:35]: will do

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+
+By the look of it, the package's reason of being is rather self-evident. `torch` itself does not -- nor should it -- implement all the newly-published, potentially-useful-for-your-purposes optimization algorithms out there.[^2] The algorithms assembled here, then, are probably exactly those the authors were most eager to experiment with in their own work. As of this writing, they comprise, amongst others, various members of the popular *ADA*\* and \**ADAM*\* families.[^3] And we may safely assume the list will grow over time.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2022-05-18-torchopt/torchopt.Rmd:37]: hm, for the readership, I think these will be familiar... the star is being used as in the unix shell (not as a formal regex)

File: _posts/2022-05-18-torchopt/torchopt.Rmd (hunk 4 above)
Comment (skeydan) [reply _posts/2022-05-18-torchopt/torchopt.Rmd:155]: yay :-)

File: _posts/2022-05-18-torchopt/torchopt.Rmd (hunk 5 above)
Comment (skeydan) [reply _posts/2022-05-18-torchopt/torchopt.Rmd:185]: thank you!

File: _posts/2022-05-18-torchopt/torchopt.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2022-05-18-torchopt/torchopt.Rmd:68]: true that, too

**Comment by skeydan** [comment]
Thanks a lot @batpigandme , super instructive as always!!

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+```{r}
+flower <- function(x, y) {
+  a <- 1
+  b <- 1
+  c <- 4
+  a * torch_sqrt(torch_square(x) + torch_square(y)) + b * torch_sin(c * torch_atan2(y, x))
+}
+```
+
+Let's begin with our explorations.
```
Comment (dfalbel) [inline _posts/2022-05-18-torchopt/torchopt.Rmd:68]: Would it make sense to show a plot of the flower function here, and maybe explain a little how to read that plot?

File: _posts/2022-05-18-torchopt/torchopt.Rmd [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+While it's true that I have no intent of comparing (let alone analyzing) different strategies, there is one that, to me, stands out in the list: ADAHESSIAN [@abs-2006-00719], a second-order algorithm designed to scale to large neural networks. I'm especially curious to see how it behaves as compared to L-BFGS, the second-order "classic" available from base `torch` we've had a [dedicated blog post](https://blogs.rstudio.com/ai/posts/2021-04-22-torch-for-optimization/) about last year.
+
+## The way it works
+
+The utility function in question is named `test_optim()`. The only required argument concerns the optimizer to try (`optim`). But you'll likely want to tweak three others as well:
+
+-   `test_fn`: To use a test function different from the default (`beale`). You can choose among the many provided in `torchopt`, or you can pass in your own. In the latter case, you also need to provide information about search domain and starting points. (We'll see that in an instant.)
+-   `steps`: To set the number of optimization steps.
+-   `opt_hparams`: To modify optimizer hyperparameters; most notably, the learning rate.
+
```
Comment (dfalbel) [inline _posts/2022-05-18-torchopt/torchopt.Rmd:54]: Maybe mention it can also produce the GIF's?

**Review by dfalbel — APPROVED** [review]
Very nice blogpost @skeydan !
Looks great to me :)

File: _posts/2022-05-18-torchopt/torchopt.Rmd (hunk 12 above)
Comment (skeydan) [reply _posts/2022-05-18-torchopt/torchopt.Rmd:68]: Yeah, I can see what you mean... I was intentionally not doing it, because of the many plots  to come ...but then, it doesn't hurt, really ;-))
And true, makes sense to shortly explain the plot, especially since - as I noticed - the color coding works exactly the opposite way from what I was doing for the other post!

File: _posts/2022-05-18-torchopt/torchopt.Rmd (hunk 13 above)
Comment (skeydan) [reply _posts/2022-05-18-torchopt/torchopt.Rmd:54]: honestly, my experiences with those weren't so good (wild flickering, etc.)
I think I'll forego that :-)

**Comment by skeydan** [comment]
Thanks a lot @dfalbel, will add the plot and the explanation!
